File: iOS/CA2D/Graphics/WorldModel.swift
//import OpenGLES
//import GLKit
//
//final class WorldModel {
//    let world: World
//
//    var vertices: [Vertex] {
//        let cellWidth = 2.0 / Float(world.width)
//        let cellHeight = 2.0 / Float(world.height)
//
//        var vertices: [Vertex] = []
//
//        for y in (0..<world.height) {
//            for x in (0..<world.width) {
//                let condition = world.cells[y * world.width + x]
//                if condition == 0 {
//                    continue
//                }
//
//                let color = colorFromCondition(condition)
//
//                let posA = GLKVector2Make(-1.0 + cellWidth * Float(x),     -1.0 + cellHeight * Float(y))
//                let posB = GLKVector2Make(-1.0 + cellWidth * Float(x),     -1.0 + cellHeight * Float(y + 1))
//                let posC = GLKVector2Make(-1.0 + cellWidth * Float(x + 1), -1.0 + cellHeight * Float(y))
//                let posD = GLKVector2Make(-1.0 + cellWidth * Float(x + 1), -1.0 + cellHeight * Float(y + 1))
//
//                vertices.appendContentsOf([
//                    Vertex(position: posA, color: color),
//                    Vertex(position: posB, color: color),
//                    Vertex(position: posC, color: color),
//
//                    Vertex(position: posB, color: color),
//                    Vertex(position: posC, color: color),
//                    Vertex(position: posD, color: color),
//                ])
//            }
//        }
//
//        return vertices
//    }
//
//    init(world: World) {
//        self.world = world
//    }
//
//    func colorFromCondition(condition: Int) -> GLKVector3 {
//        guard world.rule.conditions > 2 else {
//            return GLKVector3Make(1.0, 1.0, 0.0)
//        }
//
//        let cyan:Float = 0.0
//        let magenta:Float = 1.0 - (1.0 / Float(world.rule.conditions - 2) * Float(condition - 1))
//        let yellow:Float = 1.0
//        let key:Float = 0.0
//
//        let red   = 1 - min(1.0, cyan    * (1.0 - key)) + key
//        let green = 1 - min(1.0, magenta * (1.0 - key)) + key
//        let blue  = 1 - min(1.0, yellow  * (1.0 - key)) + key
//
//        return GLKVector3Make(red, green, blue)
//    }
//}

File: Android/app/src/main/kotlin/cc/clv/android/ca2d/graphics/WorldModel.kt
package cc.clv.android.ca2d.graphics

import cc.clv.android.ca2d.World
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.nio.FloatBuffer

class WorldModel(world: World) {
    companion object {
        val PositionSize = 4 * 2
        val ColorSize = 4 * 3
    }

    private val world = world
    private val maxVertexCount: Int
        get() = world.width * world.height * 6

    var vertexCount: Int = 0
    var positionBuffer: FloatBuffer
    var colorBuffer: FloatBuffer

    init {
        positionBuffer = ByteBuffer.allocateDirect(maxVertexCount * PositionSize)
                .order(ByteOrder.nativeOrder()).asFloatBuffer()
        colorBuffer = ByteBuffer.allocateDirect(maxVertexCount * ColorSize)
                .order(ByteOrder.nativeOrder()).asFloatBuffer()
    }

    fun update() {
        positionBuffer.clear()
        colorBuffer.clear()
        vertexCount = 0

        val cellWidth = 2.0f / world.width.toFloat()
        val cellHeight = 2.0f / world.height.toFloat()
        val colors = (0..world.rule.conditions).map { colorFromCondition(it) }

        for (y in (0..(world.height - 1))) {
            for (x in (0..(world.width - 1))) {
                val condition = world.cells[y * world.width + x]

                if (condition == 0.toByte()) {
                    continue
                }

                val posA = floatArrayOf(-1.0f + cellWidth * x.toFloat(), -1.0f + cellHeight * y.toFloat())
                val posB = floatArrayOf(-1.0f + cellWidth * x.toFloat(), -1.0f + cellHeight * (y + 1).toFloat())
                val posC = floatArrayOf(-1.0f + cellWidth * (x + 1).toFloat(), -1.0f + cellHeight * y.toFloat())
                val posD = floatArrayOf(-1.0f + cellWidth * (x + 1).toFloat(), -1.0f + cellHeight * (y + 1).toFloat())

                val color = colors[condition.toInt()]

                positionBuffer.put(posA)
                colorBuffer.put(color)

                positionBuffer.put(posB)
                colorBuffer.put(color)

                positionBuffer.put(posC)
                colorBuffer.put(color)

                positionBuffer.put(posB)
                colorBuffer.put(color)

                positionBuffer.put(posC)
                colorBuffer.put(color)

                positionBuffer.put(posD)
                colorBuffer.put(color)

                vertexCount += 6
            }
        }
    }

    fun colorFromCondition(condition: Int): FloatArray {
        if (world.rule.conditions <= 2) {
            return floatArrayOf(1.0f, 1.0f, 0.0f)
        }

        val cyan = 0.0f
        val magenta = 1.0f - (1.0f / (world.rule.conditions - 2).toFloat() * (condition - 1).toFloat())
        val yellow = 1.0f
        val key = 0.0f

        val red = 1.0f - arrayOf(1.0f, cyan * (1.0f - key)).min()!! + key
        val green = 1.0f - arrayOf(1.0f, magenta * (1.0f - key)).min()!! + key
        val blue = 1.0f - arrayOf(1.0f, yellow * (1.0f - key)).min()!! + key

        return floatArrayOf(red, green, blue)
    }
}
